### PR TITLE
Fix memory usage when listing pods/events

### PR DESCRIPTION
**Pull request by marseel**
**What type of PR is this?**
Limits memory usage after slo-monitor restart

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: slo-monitor/src/monitors/pod_monitor.go [hunk 1]
```diff
@@ -130,7 +130,15 @@ func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
 				// This call is paginated by cache.Reflector.
-				return pm.kubeClient.CoreV1().Pods(v1.NamespaceAll).List(options)
+				o := metav1.ListOptions{
```
Comment (wojtek-t) [inline slo-monitor/src/monitors/pod_monitor.go:133]: Add a comment explaining why we're doing this strange thing (also remove the comment above, it's no longer really relevant).

File: slo-monitor/src/monitors/pod_monitor.go [hunk 2]
```diff
@@ -130,7 +130,15 @@ func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
...
+					Limit: 1,
+				}
+				result, err := pm.kubeClient.CoreV1().Pods(v1.NamespaceAll).List(o)
+				if err != nil {
+					return nil, err
+				}
+				result.Continue = ""
```
Comment (wojtek-t) [inline slo-monitor/src/monitors/pod_monitor.go:140]: Let's also remove the Items of the list - we don't really list them - it will be easier to catch errors if someone will change other parts.

File: slo-monitor/src/monitors/pod_monitor.go [hunk 3]
```diff
@@ -175,9 +183,16 @@ func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
 	eventController := NewWatcherWithHandler(
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				options.FieldSelector = eventSelector
-				// This call is paginated by cache.Reflector.
-				return pm.kubeClient.CoreV1().Events(v1.NamespaceAll).List(options)
+				o := metav1.ListOptions{
```
Comment (wojtek-t) [inline slo-monitor/src/monitors/pod_monitor.go:186]: Same comments as above...

File: slo-monitor/src/monitors/pod_monitor.go [hunk 4]
```diff
@@ -175,9 +183,16 @@ func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
...
+					Limit: 1,
+					FieldSelector: eventSelector,
```
Comment (wojtek-t) [inline slo-monitor/src/monitors/pod_monitor.go:188]: Actually - you don't want this FieldSelector.

File: slo-monitor/src/monitors/watcher.go [hunk 5]
```diff
@@ -28,7 +28,6 @@ import (
 
 const (
 	resyncPeriod      = time.Duration(0)
-	watchListPageSize = 5000
```
Comment (wojtek-t) [inline slo-monitor/src/monitors/watcher.go:31]: @mborsz - given we don't need this, do we want to revert the changes you were doing?

Comment (mborsz) [reply slo-monitor/src/monitors/watcher.go:31]: Yes, once we validate this approach, we can submit follow up PR that will remove this, fork of client-go and other ugly stuff I added.

File: slo-monitor/src/monitors/pod_monitor.go (hunk 2 above)
Comment (marseel) [reply slo-monitor/src/monitors/pod_monitor.go:140]: Done.

File: slo-monitor/src/monitors/pod_monitor.go (hunk 4 above)
Comment (marseel) [reply slo-monitor/src/monitors/pod_monitor.go:188]: Done.

File: slo-monitor/Makefile [hunk 6]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 PACKAGE = k8s.io/perf-tests/slo-monitor
-TAG = 0.14.0
+TAG = 0.14.1
```
Comment (wojtek-t) [inline slo-monitor/Makefile:16]: let's make 0.15 - it's a bit non-negligible change

File: slo-monitor/src/monitors/pod_monitor.go (hunk 1 above)
Comment (marseel) [reply slo-monitor/src/monitors/pod_monitor.go:133]: Done.

File: slo-monitor/src/monitors/pod_monitor.go (hunk 3 above)
Comment (marseel) [reply slo-monitor/src/monitors/pod_monitor.go:186]: Done.

File: slo-monitor/src/monitors/pod_monitor.go (hunk 4 above)
Comment (marseel) [reply slo-monitor/src/monitors/pod_monitor.go:188]: Done.

File: slo-monitor/Makefile (hunk 6 above)
Comment (marseel) [reply slo-monitor/Makefile]: Done.

File: slo-monitor/src/monitors/pod_monitor.go [hunk 7]
```diff
@@ -175,9 +186,19 @@ func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
 	eventController := NewWatcherWithHandler(
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				options.FieldSelector = eventSelector
-				// This call is paginated by cache.Reflector.
-				return pm.kubeClient.CoreV1().Events(v1.NamespaceAll).List(options)
+				// slo monitor is not using list of events.
```
Comment (mborsz) [inline slo-monitor/src/monitors/pod_monitor.go:189]: I would add more context, like: NewWatcherWithHandler internally uses NoStore store that ignores all relist results, so here we do List call only to get the current resource version.

File: slo-monitor/src/monitors/pod_monitor.go [hunk 8]
```diff
@@ -129,8 +129,19 @@ func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) error {
 	controller := NewWatcherWithHandler(
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				// This call is paginated by cache.Reflector.
-				return pm.kubeClient.CoreV1().Pods(v1.NamespaceAll).List(options)
+				// slo monitor is not using list of pods.
+				// It's only used to get resourceVersion for Watch.
+				// This implementation allows us to reduce memory usage significantly for large clusters.
```
Comment (wojtek-t) [inline slo-monitor/src/monitors/pod_monitor.go:134]: nit: 
``` 
slo-monitor is not using the pods returned from the list
(it uses NoStore implementation of Store which discards them),
only the resourceVersion is used to instantiate the watch from this point.
This trick allows us to reduce memory usage on startup and further relists.
```

**Comment by wojtek-t** [comment]
/ok-to-test

/approve

Will let @mborsz to lgtm

**Comment by mborsz** [comment]
/lgtm

Thanks!
